ComponentKit/Core/CKBuildComponent.mm

```diff
@@ -17,6 +17,44 @@
 #import "CKComponentScopeRoot.h"
 #import "CKComponentSubclass.h"
 #import "CKThreadLocalComponentScope.h"
+#import <ComponentKit/CKInternalHelpers.h>
+#import "CKComponentTreeNode.h"
+
+static void CKBuildComponentTree(CKComponent *c,
+                                 CKComponentTreeNode *parent,
+                                 CKComponentTreeNode *previousParent,
+                                 CKComponentScopeRoot *scopeRoot,
+                                 const CKComponentStateUpdateMap &stateUpdates) {
+
+  // Reconciliation:
+  //
+  // This is the place that we could intgerate reconciliation:
+  // We can compare the current component with the previous one from the 'previousParent';
+  // In case that it shouldn't be recreated, we can just return the previous child node and skip the whole tree creation from this level.
+  //
+  // We still need to find a way to check if a component in the tree had a state update.
+  // One possible soltion is alternate the tree node identifier to hold a node of identifiers.
+
+  CKComponentTreeNode *node = [[CKComponentTreeNode alloc]
+                               initWithParent:parent
+                               previousParent:previousParent
+                               scopeRoot:scopeRoot
+                               component:c
+                               stateUpdates:stateUpdates];
+
+  auto const children = [c render:node.state];
+
+  // In case of that the render method returns one child, we will set it on the parent component in order to take it into account during layout time.
+  // A better solution would be to unify the render callbacks with the layout process and have a direct access to the component children.
+  if (children.size() == 1) {
+    c.childComponent = children[0];
```

Comment:
What happens if `CKBuildComponentTree` is called simultaneously on two different threads? For example, this happens quite frequently when rendering is performed on the main thread and a background thread in parallel.

Reply:
`CKBuildComponentTree` is being called only once during the component creation, it cannot be call twice. However, you are right that if we are going to implement reconciliation and reuse a component tree node - it won't work. This was a first quick solution for the RFC.

An ideal solution would be to access the tree node from the layout computation. That means that the component itself can be attached to a two different nodes at any given time and it doesn't matter as it's immutable. The information about the children is being encapsulated in the tree node.